sstp_protocol/src/parse.rs

```diff
@@ -0,0 +1,986 @@
+            if &buffer == b"Conflict\r\n" {
+                Ok(StatusCode::CONFLICT)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"420 " => {
+            let buffer = read!(cursor, 8).map_err(Error::from)?;
+            if &buffer == b"Refuse\r\n" {
+                Ok(StatusCode::REFUSE)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"501 " => {
+            let buffer = read!(cursor, 17).map_err(Error::from)?;
+            if &buffer == b"Not Implemented\r\n" {
+                Ok(StatusCode::NOT_IMPLEMENTED)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"503 " => {
+            let buffer = read!(cursor, 21).map_err(Error::from)?;
+            if &buffer == b"Service Unavailable\r\n" {
+                Ok(StatusCode::SERVICE_UNAVAILABLE)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"510 " => {
+            let buffer = read!(cursor, 14).map_err(Error::from)?;
+            if &buffer == b"Not Local IP\r\n" {
+                Ok(StatusCode::NOT_LOCAL_IP)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"511 " => {
+            let buffer = read!(cursor, 15).map_err(Error::from)?;
+            if &buffer == b"In Black List\r\n" {
+                Ok(StatusCode::IN_BLACK_LIST)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        b"512 " => {
+            let buffer = read!(cursor, 11).map_err(Error::from)?;
+            if &buffer == b"Invisible\r\n" {
+                Ok(StatusCode::INVISIBLE)
+            } else {
+                Err(Error::InvalidStatusCode)
+            }
+        }
+        _ => Err(Error::InvalidStatusCode),
+    }
+}
+
+fn parse_headers(cursor: &mut Cursor<&[u8]>) -> Result<HeaderMap> {
+    let mut map = HeaderMap::new();
+    loop {
+        let buffer = lookahead!(cursor, 2).map_err(Error::from)?;
+        if &buffer == b"\r\n" {
+            let _ = read!(cursor, 2);
+            break;
+        }
+
+        let name = read_until!(cursor, b":").map_err(Error::from)?;
+        let name = HeaderName::from_bytes(name.as_slice()).map_err(Error::from)?;
+
+        let _ = read_repeat!(cursor, b" ").map_err(Error::from)?;
+
+        let value = read_until!(cursor, b"\r\n").map_err(Error::from)?;
+
+        map.insert(name, value)
+    }
+
+    Ok(map)
+}
+
+fn parse_additional_data(cursor: &mut Cursor<&[u8]>) -> Result<AdditionalData> {
+    let result = lookahead!(cursor, 1);
+    if result.is_err() && result.unwrap_err().kind() == std::io::ErrorKind::UnexpectedEof {
+        return Ok(AdditionalData::Empty);
+    }
+    let mut bytes = Vec::new();
+    loop {
+        let buffer = lookahead!(cursor, 2).map_err(Error::from)?;
+        if &buffer == b"\r\n" {
```

Comment:
If it is to be a constant, how would you replace it with a constant in the case of this code?

```rust
    let buffer = read!(cursor, 10).map_err(Error::from)? ;
    match &buffer {
        b "SSTP/1.0\r\n" => Ok(Version::SSTP_10),
        b "SSTP/1.1\r\n" => Ok(Version::SSTP_11),
        b "SSTP/1.2\r\n" => Ok(Version::SSTP_12),
        b "SSTP/1.3\r\n" => Ok(Version::SSTP_13),
        b "SSTP/1.4\r\n" => Ok(Version::SSTP_14),
        _ => Err(Error::InvalidVersion),
    }
```

It would have been nice if we could concat with a macro, for example, but this would result in an error.

```rust
concat_bytes!(b "SSTP/1.0", NEW_LINE) => ...
```

Reply:
```rus
concat_bytes!(b "SSTP/1.0", [10, 13]) => ...
```

Yes, this works.
However, this method cannot use constants. Only literals can be used.

When using constants, I believe it is important to replace all the same values with constants.
Is there a way to use constants in all `\r\n`?

Reply:
Thank you. I do not use constants.